odl/contrib/pytorch/layer.py

```diff
@@ -0,0 +1,241 @@
+
+        Compute the gradient of a custom functional:
+
+        >>> space = odl.rn(3)
+        >>> functional = odl.solvers.L2NormSquared(space)
+        >>> func_op = TorchOperator(functional)
+        >>> x_elem = space.element([1, 2, 3])
+        >>> x = torch.from_numpy(x_elem.asarray())
+        >>> x_var = torch.autograd.Variable(x, requires_grad=True)
+        >>> func_x_var = func_op(x_var)
+        >>> func_x_var
+        Variable containing:
+         14
+        [torch.DoubleTensor of size 1]
+
+        >>> func_x_var.backward()
+        >>> x_var.grad  # Should be 2 * x
+        Variable containing:
+         2
+         4
+         6
+        [torch.DoubleTensor of size 3]
+
+        Notes
+        -----
+        This method applies the contribution of this node, i.e., the
+        transpose of the Jacobian of its outputs with respect to its inputs,
+        to the gradients of some cost function with respect to the outputs
+        of this node.
+
+        Example: Assume that this node computes :math:`x \mapsto f(x)`, and
+        its output is connected to a cost function that computes
+        :math:`y --> C(y)`. Here, :math:`x`, :math:`y` and :math:`f(x)` are
+        tensor variables and :math:`C(y)` is a scalar variable.
+        In ODL language, what ``backward`` should compute is
+
+            .. math::
+                \\nabla(C \circ f)(x) = f'(x)^*\\big(\\nabla C (f(x))\\big)
+
+        according to the chain rule. In ODL code, this corresponds to ::
+
+            f.derivative(x).adjoint(C.gradient(f(x))).
+
+        Then, the parameter ``grad_output`` contains a single tensor
+        variable ``y`` that stands for :math:`\\nabla C(f(x))`. Thus,
+        ``backward`` boils down to computing ``[f'(x)^*(y)]``, where
+        ``grad_output == [y]`` and ``inputs == [x]``.
+        """
+        if not self.operator.is_linear:
+            input, = self.saved_variables
+            input_arr = input.data.numpy()
+            if all(s == 0 for s in input_arr.strides):
+                # TODO: remove when Numpy issue #9165 is fixed
+                # https://github.com/numpy/numpy/pull/9177
+                input_arr = input_arr.copy()
+
+        grad = None
+
+        # ODL weights spaces, pytorch doesn't, so we need to handle this
+        try:
+            dom_weight = self.operator.domain.weighting.const
+        except AttributeError:
+            dom_weight = 1.0
+
+        try:
+            ran_weight = self.operator.range.weighting.const
+        except AttributeError:
+            ran_weight = 1.0
+
+        scaling = dom_weight / ran_weight
+
+        if self.needs_input_grad[0]:
+            grad_output_arr = grad_output.numpy()
+            if all(s == 0 for s in grad_output_arr.strides):
+                # TODO: remove when Numpy issue #9165 is fixed
+                # https://github.com/numpy/numpy/pull/9177
+                grad_output_arr = grad_output_arr.copy()
+
+            if self.operator.is_linear:
```

Comment:
This if statement does nothing. There is already a shortcut in `operator.derivative`.

If you truly want to have a performance gain, you would put this if statement around the creation of the array `input_arr`, which is not needed.

Reply:
That's what I'm doing. The input is saved for usage here only if the operator is non-linear, so I have it only in that case and need the distinction also here.